controller/src/main/java/io/pravega/controller/store/stream/PravegaTablesStoreHelper.java

```diff
@@ -0,0 +1,403 @@
+/**
+ * Copyright (c) 2017 Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.controller.store.stream;
+
+import io.netty.buffer.ByteBuf;
+import io.pravega.client.tables.impl.IteratorState;
+import io.pravega.client.tables.impl.KeyVersion;
+import io.pravega.client.tables.impl.KeyVersionImpl;
+import io.pravega.client.tables.impl.TableEntry;
+import io.pravega.client.tables.impl.TableEntryImpl;
+import io.pravega.client.tables.impl.TableKey;
+import io.pravega.client.tables.impl.TableKeyImpl;
+import io.pravega.common.Exceptions;
+import io.pravega.common.concurrent.Futures;
+import io.pravega.common.tracing.RequestTag;
+import io.pravega.common.util.AsyncIterator;
+import io.pravega.common.util.ContinuationTokenAsyncIterator;
+import io.pravega.controller.server.SegmentHelper;
+import io.pravega.controller.server.WireCommandFailedException;
+import io.pravega.controller.server.rpc.auth.AuthHelper;
+import io.pravega.controller.store.host.HostStoreException;
+import io.pravega.controller.util.RetryHelper;
+import lombok.EqualsAndHashCode;
+import lombok.NonNull;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.lang3.tuple.ImmutablePair;
+import org.apache.commons.lang3.tuple.Pair;
+import org.apache.curator.shaded.com.google.common.base.Charsets;
+
+import java.util.AbstractMap;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+
+import static io.pravega.controller.store.stream.AbstractStreamMetadataStore.DATA_NOT_FOUND_PREDICATE;
+
+@Slf4j
+public class PravegaTablesStoreHelper {
```

Comment:
My main concern with this class is that I can't tell whether we are handling errors coming from the segment store appropriately. Without a spec, I can't really verify. 

In the past, we have been bitten a few times due to mishandling errors from the segment store, and it would be wise to look carefully into it.

Reply:
The class that deals with segment store is `SegmentHelper`. 
Any issues with mishandling of responses from segmentStore should be done there. 
This is an internal wrapper class whose contract is with segmenthelper. 
It handles all "expected" failure responses from SegmentHelper correctly. 

If there is an issue of unhandled response from segment store, we should be handling it there. 

So for all table related apis that were added to segmentHelper, the request and responses were both added by @shrids and he has added unit tests for them as well in segmenthelperTest.